error when registering meta kernels to composite ops in core

bdhirsh opened the pull request:
More detailed description in the code. We shouldn't register meta kernels directly to ops that have CompositeImplicitAutograd kernels.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79741

bdhirsh committed:
error when registering meta kernels to composite ops in core

[ghstack-poisoned]

Chillee commented on torch/_meta_registrations.py:
```diff
@@ -113,7 +113,9 @@ def checkUplo(uplo: str):
-@torch.library.impl(meta_lib, "linalg_eigh")
+# Keeping this meta impl around, but we don't want to register it directly to the meta key
+# because `aten::linalg_eigh` is composite.
+# `_linalg_eigh` is implemented internally as a structured kernel, so we have meta support.
```
we might need a python meta function anyways, for the purposes of symbolic shape propagation.

Chillee reviewed (approved):
I also expect that this will expose some errors (such as for the `aten.where` registration?)

bdhirsh commented:
Oh yep, there are... a bunch haha. I forgot to `import torch._decomp` when I was testing locally.

I'll fix them up as part of the PR

bdhirsh committed:
Update on "error when registering meta kernels to composite ops in core"

More detailed description in the code. We shouldn't register meta kernels directly to ops that have CompositeImplicitAutograd kernels.




[ghstack-poisoned]

bdhirsh commented:
I ended up updating the register decomposition code to just check if each overload has a CompositeImplicitAutograd kernel and skip the meta registration if it does. The list of "offenders":
```
aten::mish_backward
aten::special_i0
aten::special_i0.out
aten::bitwise_or.Scalar
aten::bitwise_xor.Scalar
aten::where.ScalarSelf
aten::where.ScalarOther
aten::where.Scalar
aten::where
aten::sum.dim_DimnameList
aten::sum.DimnameList_out
aten::prod.dim_Dimname
aten::prod.Dimname_out
aten::cat.names
aten::cat.names_out
```

I didn't carefully check to make sure that all of those op's have meta kernels for their base ops - sorta hoping CI will catch that (I can probably check quickly tomorrow though)

bdhirsh committed:
Update on "error when registering meta kernels to composite ops in core"

More detailed description in the code. We shouldn't register meta kernels directly to ops that have CompositeImplicitAutograd kernels.




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4147,7 +4147,7 @@
   dispatch:
     CPU, CUDA: mish_backward
-    CompositeImplicitAutograd: math_mish_backward
+    CompositeExplicitAutograd: math_mish_backward
```
So it is not possible to do `CompositeImplicit` + another key anymore?

ngimel replied:
Is `mish_backward` different numerically from the one that would be obtained in the decomp?

bdhirsh replied:
Just tested it out and it looks like you're right - `torch.allclose` with the defaults thinks they're the same numerically:
```
def f(a, b):
    return torch.ops.aten.mish_backward(a, b)

a = torch.randn(8, dtype=torch.float64)
b = torch.randn(8, dtype=torch.float64)

out1 = f(a, b)
out2 = functionalize(f)(a, b)

print(torch.allclose(out1, out2))

f_traced = make_fx(f)(a, b)
f_no_mutations_traced = make_fx(functionalize(f))(a, b)

print(f_traced.code)
print(f_no_mutations_traced.code)
```
prints:
```
def forward(self, a_1, b_1):
    mish_backward = torch.ops.aten.mish_backward(a_1, b_1);  a_1 = b_1 = None
    return mish_backward




def forward(self, a_1, b_1):
    softplus = torch.ops.aten.softplus(b_1)
    tanh = torch.ops.aten.tanh(softplus);  softplus = None
    sigmoid = torch.ops.aten.sigmoid(b_1)
    mul = torch.ops.aten.mul(tanh, tanh)
    empty_like = torch.ops.aten.empty_like(mul, memory_format = torch.preserve_format)
    fill = torch.ops.aten.fill(empty_like, 1);  empty_like = None
    sub = torch.ops.aten.sub(fill, mul);  fill = mul = None
    mul_1 = torch.ops.aten.mul(b_1, sigmoid);  b_1 = sigmoid = None
    mul_2 = torch.ops.aten.mul(mul_1, sub);  mul_1 = sub = None
    add = torch.ops.aten.add(tanh, mul_2);  tanh = mul_2 = None
    mul_3 = torch.ops.aten.mul(a_1, add);  a_1 = add = None
    return mul_3
```

Would you rather keep the op as `CompositeImplicitAutograd`? I guess the pros/cons are pretty minor:
- pro: the output of your trace will be the same regardless of whether you're running a transform
- con: none? (since CPU/CUDA will treat it as a primitive today, so only e.g. XLA / functorch transforms will actually use the composite.

bdhirsh committed:
Update on "error when registering meta kernels to composite ops in core"

More detailed description in the code. We shouldn't register meta kernels directly to ops that have CompositeImplicitAutograd kernels.




[ghstack-poisoned]

bdhirsh committed:
Update on "error when registering meta kernels to composite ops in core"

More detailed description in the code. We shouldn't register meta kernels directly to ops that have CompositeImplicitAutograd kernels.




[ghstack-poisoned]

eellison commented:
@pytorchbot rebase

pytorchmergebot committed:
Update on "error when registering meta kernels to composite ops in core"

More detailed description in the code. We shouldn't register meta kernels directly to ops that have CompositeImplicitAutograd kernels.




[ghstack-poisoned]

eellison commented:
@pytorchbot merge
